File: src/main/java/com/zepben/vertxutils/routing/ChunkedResponse/HttpChunkedJsonResponse.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.ChunkedResponse;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import io.vertx.core.http.HttpServerResponse;

@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public class HttpChunkedJsonResponse extends ChunkedJsonResponse {

    private final HttpServerResponse response;
    private final int bufferSize;

    public HttpChunkedJsonResponse(HttpServerResponse response) {
        this(response, DEFAULT_BUFFER_SIZE);
    }

    public HttpChunkedJsonResponse(HttpServerResponse response, int bufferSize) {
        super(bufferSize);

        this.response = response;
        this.bufferSize = bufferSize;
    }

    @Override
    protected void end(StringBuilder sb) {
        if (!response.closed()){
            response.end(sb.toString());
        }
    }

    @Override
    protected void send(boolean force, StringBuilder sb) {
        if ((force || sb.length() >= bufferSize) && !response.closed()) {
            response.write(sb.toString());
            sb.setLength(0);
        }
    }

}

File: src/main/java/com/zepben/vertxutils/routing/chunked/HttpChunkedJsonResponse.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing.chunked

import io.netty.handler.codec.http.HttpResponseStatus
import io.vertx.core.http.HttpServerResponse

class HttpChunkedJsonResponse(
    private val response: HttpServerResponse,
    private val bufferSize: Int = DEFAULT_BUFFER_SIZE,
) : ChunkedJsonResponse(bufferSize) {

    private var canSetStatus = true

    override fun onResponseCompleted(sb: StringBuilder) {
        canSetStatus = false
        if (!response.closed()) {
            response.end(sb.toString())
        }
    }

    override fun checkWrite(sb: StringBuilder, force: Boolean) {
        canSetStatus = false
        if ((force || (sb.length >= bufferSize)) && !response.closed()) {
            response.write(sb.toString())
            sb.setLength(0)
        }
    }

    var statusCode: HttpResponseStatus
        get() = HttpResponseStatus.valueOf(response.statusCode)
        set(status) {
            // Once a response is committed (first write or end), it is too late to change the status.
            check(canSetStatus) { "You can't set the status after the response has been committed" }
            response.statusCode = status.code()
        }

}
